### DISCUSSION TITLE
Not found internal modules anymore

### DISCUSSION THREAD

**Discussion by Gotiap** [oqtane/oqtane.framework#1330] [Q&A]
Hello,

As I can see now we do not have the option available to create an internal module since version 2.0.2, was available in version 2.0.1. Is there any alternate way to create the Internal Module?

**Comment by leigh-pointer** [comment]
Hi @Gotiap 
Internal Module Creator was removed as it is an anti-pattern. 
To quote @sbwalker 

> the Internal Module Creator was removed as it is an anti-pattern. The whole point of a modular framework is to extend it with new functionality without modifying the core framework. The Internal Module Creator was adding code to the core framework which breaks this rule.

[https://github.com/oqtane/oqtane.framework/discussions/1256](https://github.com/oqtane/oqtane.framework/discussions/1256)

**Reply by Gotiap** [reply]
Thank you.

**Reply by jasza2** [reply]
> Hi @Gotiap
> Internal Module Creator was removed as it is an anti-pattern.
> To quote @sbwalker
> 
> > the Internal Module Creator was removed as it is an anti-pattern. The whole point of a modular framework is to extend it with new functionality without modifying the core framework. The Internal Module Creator was adding code to the core framework which breaks this rule.
> 
> #1256

Pity because development is slowed down a lot by having to attach to the process every time to debug , this would be OK on its own but you also have to stop and start the site every time. Is there a way we can add the option back in? It would be good to break the back of the module development internally and then copy the core code to an external module when ready to package for external use. Most of our developmemt is for internal use and would never need to be packaged so we would be taking an estimated 30 percent productivity hit on daily development for something we only need to do once in a while

**Reply by leigh-pointer** [reply]
> Pity because development is slowed down a lot by having to attach to the process every time to debug
@jasza2 @Gotiap 
**Try this.**
Create your project (external module) add the Oqtane Server project to the solution (no need to reference it, just add it) .  Set the server project as startup.
Debug away.
I discovered this method purely by accident and has worked well ever since.

**Reply by jasza2** [reply]
Thanks for the quick reply this works well. However there is no formatting to the website it's all plain text. Any tips on getting back the good looks.?

**Reply by leigh-pointer** [reply]
@jasza2 I always build the module for release first to make sure the module content is in the wwwroot folder.  If you have done that, make sure your styles sheets are registered correctly.

**Comment by sbwalker** [comment]
In 2.1.0 the ability to create Internal modules was added back to the Module Creator - however the template was not included so you would need to create your own Internal module template in the Oqtane.Server/wwwroot/Modules/Templates folder (you should be able to use the template code files from version 2.0.0). There is also a new template.json where you can include metadata for your template - and you would need to set the Type to "Internal".
